maestro-android/src/main/java/dev/mobile/maestro/receivers/LocaleSettingReceiver.kt

```diff
@@ -0,0 +1,105 @@
+import android.content.Context
+import android.content.Intent
+import android.util.Log
+import dev.mobile.maestro.handlers.LocaleSettingHandler
+import org.apache.commons.lang3.LocaleUtils
+import java.util.*
+
+class LocaleSettingReceiver : BroadcastReceiver(), HasAction {
+    override fun onReceive(context: Context?, intent: Intent?) {
+        var language = intent?.getStringExtra(LANG)
+        var country = intent?.getStringExtra(COUNTRY)
+
+        if (language == null || country == null) {
+            Log.w(TAG, "It is required to provide both language and country, for example: " +
+                    "am broadcast -a dev.mobile.maestro --es lang ja --es country JP")
+            Log.i(TAG, "Set en-US by default.")
+            language = "en"
+            country = "US"
+        }
+
+        var locale = Locale(language, country)
+
+        Log.i(TAG, "Obtained locale: $locale")
+
+        val script = intent?.getStringExtra(SCRIPT)
+        if (script != null) {
+            Locale.Builder().setLocale(locale).setScript(script).build().also { locale = it }
+        }
+
+        if (!LocaleUtils.isAvailableLocale(locale)) {
+            val approximateMatchesLc = matchLocales(language, country)
+
+            if (approximateMatchesLc.isNotEmpty() && script.isNullOrBlank()) {
+                Log.i(
+                    TAG,
+                    "The locale $locale is not known. Selecting the closest known one ${approximateMatchesLc[0]} instead"
+                )
+                locale = approximateMatchesLc[0]
+            } else {
+                val approximateMatchesL = matchLocales(language)
+                if (approximateMatchesL.isEmpty()) {
+                    Log.e(
+                        TAG,
+                        "The locale $locale is not known. Only the following locales are available: ${LocaleUtils.availableLocaleList()}"
+                    )
+                } else {
+                    Log.e(
+                        TAG,
+                        "The locale $locale is not known. " +
+                                "The following locales are available for the $language language: $approximateMatchesL" +
+                                "The following locales are available altogether: ${LocaleUtils.availableLocaleList()}"
+                    )
+                }
+                resultCode = Activity.RESULT_CANCELED
+                resultData = "The locale $locale is not known"
+                return
+            }
+        }
+
+        context?.let {
+            LocaleSettingHandler(it).setLocale(locale)
+            Log.i(TAG, "Set locale: $locale")
+            resultCode = Activity.RESULT_OK
+            resultData = locale.toString()
+        } ?: Log.e(TAG, "Failed to set device locale setting (context is null)")
+    }
+
+    private fun matchLocales(language: String): List<Locale> {
+        val matches = ArrayList<Locale>()
+        for (locale in LocaleUtils.availableLocaleList()) {
```

Comment:
What happens when locale is not available?

Reply:
it sets `resultCode` to `Activity.RESULT_CANCELED` so we can also handle it on maestro side. there is also some logic to find the closest possible locale if one is not available.

Reply:
I'm not sure if I follow this why do we have Activity result codes in headless environment. I think we need to return some other result codes